contributors/design-proposals/containerized-mounter-pod.md

```diff
@@ -0,0 +1,200 @@
+
+To sum it up, it's just a daemon set that spawns privileged pods, running a simple init + a daemon that executes mount utilities as requested by kubelet via gRPC.
+
+## Design
+
+### Configuration of the host OS
+With the [HostPath volume propagation](propagation.md) implemented, we must ensure that `/var/lib/kubelet` is share-able into containers so a pod with mount utilities can mount something there in its container and this mount will be visible to kubelet and docker (or other container engine) on the host.
+We propose:
+*  During startup. kubelet checks that `/var/lib/kubelet` on the host is on a mount with shared mount propagation.
+    *  If not, kubelet makes `/var/lib/kubelet` shareable:
+       ```shell
+       mount --bind /var/lib/kubelet /var/lib/kubelet
+       mount --make-shared /var/lib/kubelet
+       ```
+       Note that these commands are executed on the host in case kubelet runs in a container.
+       If this startup adjustment fails, kubelet refuses to start with a clear error message.
+* Kubelet does not do any other special check. We assume that kubelet either runs on the host and in the same mount namespace as docker and the host, or it runs in a container that sees /var/lib/kubelet as it is on the host and with shared mount propagation.
+
+This ensures that kubelet runs out of the box on any distro without any configuration done by the cluster admin.
+
+### Volume plugins
+* All volume plugins need to be updated to use a new `mount.Exec` interface to call external utilities like `mount`, `mkfs`, `rbd lock` and such. Implementation of the interface will be provided by caller and will lead either to simple `os.exec` on the host or a gRPC call to a socket in `/var/lib/kubelet/plugin-sockets/` directory.
+
+### Controllers
+TODO: how will controller-manager talk to a remote pod? It's relatively easy to do something like `kubectl exec <mount pod>` from controller-manager, however it's harder to *discover* the right pod.
+
+### Kubelet
+* When kubelet talks to a volume plugin, it looks for a socket named `/var/lib/kubelet/plugin-sockets/<plugin-name>`. This allows for easier discovery of flex volume drivers - probe in https://github.com/kubernetes/community/pull/833 needs to scan `/var/lib/kubelet/plugin-sockets/` too and find sockets in any new subdirectories.
+  * If the socket does not exist, kubelet gives the volume plugin plain `os.Exec` as implementation of `mount.Exec` interface and all mount utilities are executed on the host.
+  * If the socket exists, kubelet gives the volume plugin `GRPCExec` as implementation of `mount.Exec` and all mount utilities are executed via gRPC on the socket which presumably leads to a pod with mount utilities running a gRPC server.
+
+As consequence, kubelet will try to run mount utilities on the host when it starts and has not received pods with mount utilities yet (and thus `/var/lib/kubelet/plugin-sockets/` is empty). This is likely to fails with a cryptic error:
```

Comment:
I agree that the error message looks ugly, however it's quite hard to give a better one. Kubelet does not have any configuration regarding mount pods, so it does not know in advance which volume plugins should execute their stuff on the host and which should wait for pods with mount utilities (and kubelet should therefore report a nice error that it's waiting for a pod).

Taking Gluster as an example, its volume plugin just calls `/bin/mount -t glusterfs <what> <where>`. `/bin/mount` returns universal exit code 32, which can mean many things and missing `/sbin/mount.glusterfs` is just one of them. Another one can be missing fuse kernel module. IMO, it's impossible to distinguish if the error is caused by a tool that could be provided by a pod (and thus we could report "please install a daemonset with GlusterFS utilities") or something else is broken.

What I could do is to 1) check for presence of `/sbin/mount.glusterfs` in the volume plugin and 2) scan stderr for some known strings like "unknown filesystem type 'glusterfs'". If one of these happens, mount plugin returns an error "Missing mount utilities for GlusterFS" and kubelet sends it as event on the pod.  Of course, all plugins need to be updated like this, Gluster is just an example. This is not 100% heuristics, but better than nothing.


What's worst, there *can* be mount.glusterfs on the host, but system admin wants to use the one from a pod, because the pod can have a newer version or some important bug fixed or the right file in pod's /etc. And here the mount can just fail with any error message.